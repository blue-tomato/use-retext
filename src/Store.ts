type KeysOfType<T, TProp> = { [P in keyof T]: T[P] extends TProp ? P : never }[keyof T];

export type Action<S> =
  | {
      [P in KeysOfType<S, object>]?: Action<S[P]>;
    }
  | {
      [key: string]: 0;
    };

export type ReducerFuntion = (state: any, payload?: any) => object;

export interface Reducer {
  [key: string]: Reducer | ReducerFuntion;
}

export interface Store<S> {
  state: S;
  action: Action<S>;
  reducer: Reducer;
}
