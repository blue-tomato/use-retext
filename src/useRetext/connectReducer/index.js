import { mapValuesDeep } from '../../helpers';

export default (reducer, emitter, onCall) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
