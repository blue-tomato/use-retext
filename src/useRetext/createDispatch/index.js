import mitt from 'mitt';
import { mapValuesDeep } from '../../helpers';

export default action => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => payload => {
    emitter.emit(key, { scope, payload });
  });

  return [dispatch, emitter];
};
