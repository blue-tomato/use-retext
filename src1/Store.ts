export type Actions = Record<string, 0>;

export type ReducerFunction = (state: any, payload?: any) => object;

export type Reducer = Record<string, ReducerFunction>;

export interface Store {
  [key: string]: Store | unknown;
  _actions?: Actions;
  _reducer?: Reducer;
}
