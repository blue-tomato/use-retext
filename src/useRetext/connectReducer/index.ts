import { Emitter } from 'mitt';
import { mapValuesDeep } from '../../helpers';
import { Reducer, ReducerFuntion } from '../../Store';

interface OnCallObject {
  key: string;
  scope: string;
  reducer: ReducerFuntion;
  payload: any;
}

export default (reducer: Reducer, emitter: Emitter, onCall: (event: OnCallObject) => void) => {
  mapValuesDeep(reducer, ({ key, value, scope: reducerScope }) => {
    emitter.on(key, ({ scope: emitterScope, payload }) => {
      // Ensure that the reducer-scope starts at least with the emitter-scope
      if (!reducerScope.startsWith(emitterScope)) return;

      onCall({ key, scope: reducerScope, reducer: value, payload });
    });
  });
};
