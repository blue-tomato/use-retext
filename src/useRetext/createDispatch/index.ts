import mitt from 'mitt';
import { mapValuesDeep } from '../../helpers';
import { Action } from '../../Store';

export default <S>(action: Action<S>) => {
  const emitter = mitt();
  const dispatch = mapValuesDeep(action, ({ key, scope }) => (payload: any) => {
    emitter.emit(key, { scope, payload });
  });

  return [dispatch, emitter];
};
