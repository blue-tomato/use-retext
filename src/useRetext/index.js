import { useMemo, useState } from 'react';
import { get } from 'lodash-es';
import { assert } from '../helpers';
import createDisptach from './createDispatch';
import connectReducer from './connectReducer';

const mapReducer = (reducer, setState) =>
  Object.entries(reducer).reduce(
    (acc, [key, value]) => ({
      ...acc,
      [key]:
        typeof value === 'object' ? mapReducer(value, setState) : payload => setState(state => value(state, payload)),
    }),
    {},
  );

export default store => {
  assert(typeof store !== 'object', 'Store is not an object');

  const { state: initialState, action, reducer } = store;

  assert(
    typeof initialState !== 'object' || typeof action !== 'object' || typeof reducer !== 'object',
    'State, action or reducer are not an object',
  );

  const [state, setState] = useState(initialState);
  const [disptach, emitter] = createDisptach(action);

  connectReducer(reducer, emitter, ({ scope, reducer: reducerFunction, payload }) => {
    // reducerFunction(get(state, scope), payload);
  });

  const dispatch = useMemo(() => mapReducer(reducer, setState), [reducer]);

  return { state, dispatch };
};
