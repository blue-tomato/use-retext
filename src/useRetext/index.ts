import { useState } from 'react';
import get from 'just-safe-get';
import set from 'just-safe-set';
import { assert } from '../helpers';
import createDispatch from './createDispatch';
import connectReducer from './connectReducer';
import { Store } from '../Store';

export default (store: Store) => {
  assert(typeof store !== 'object', 'Store is not an object');

  const { state: initialState, action, reducer } = store;

  assert(
    typeof initialState !== 'object' || typeof action !== 'object' || typeof reducer !== 'object',
    'State, action or reducer are not an object',
  );

  const [state, setState] = useState(initialState);
  const [dispatch, emitter] = createDispatch(action);

  connectReducer(reducer, emitter, ({ scope, reducer: reducerFunction, payload }) => {
    setState((currentState: any) => {
      const scopedState = scope ? get(currentState, scope) : currentState;
      const result = reducerFunction(scopedState, payload);
      let newState = { ...currentState };

      if (scope) {
        set(newState, scope, { ...scopedState, ...result });
      } else {
        newState = { ...scopedState, ...result };
      }

      return newState;
    });
  });

  return { state, dispatch };
};
