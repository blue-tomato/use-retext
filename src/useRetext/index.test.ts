import { renderHook, act } from '@testing-library/react-hooks';
import { Store } from '../Store';
import useRetext from '.';

const getHook = <S>(store: Store<S>) => renderHook(() => useRetext(store));
const { store } = global;

describe('store argument', () => {
  it('throws an error when no object is passed', () => {
    const { result } = getHook(store);

    expect(result.error).toMatchSnapshot();
  });

  it('throws an error when no state is passed', () => {
    const { result } = getHook({ action: {}, reducer: {} } as Store<{}>);

    expect(result.error).toMatchSnapshot();
  });

  it('throws an error when no reducer is passed', () => {
    const { result } = getHook({ state: {}, action: {} } as Store<{}>);

    expect(result.error).toMatchSnapshot();
  });

  it('throws an error when no action is passed', () => {
    const { result } = getHook({ state: {}, reducer: {} } as Store<{}>);

    expect(result.error).toMatchSnapshot();
  });
});

describe('return', () => {
  it('keeps the initial state', () => {
    const { result } = getHook(store);
    const { state } = result.current;

    expect(state).toMatchSnapshot();
  });

  it('creates a disptach object', () => {
    const { result } = getHook(store);
    const { dispatch } = result.current;

    expect(dispatch).toMatchSnapshot();
  });
});

describe('in application usage', () => {
  it('increments state', () => {
    const { result } = getHook(store);

    act(() => {
      result.current.dispatch.increment();
      result.current.dispatch.sideMenu.increment();
    });

    expect(result.current.state.count).toEqual(1);
    expect(result.current.state.sideMenu.maxItems).toEqual(102);
  });

  it('toggles state', () => {
    const { result } = getHook(store);

    act(() => {
      result.current.dispatch.sideMenu.toggle();
    });

    expect(result.current.state.sideMenu.isOpen).toEqual(true);
    expect(result.current.state.sideMenu.child.isExpanded).toEqual(true);
  });
});
