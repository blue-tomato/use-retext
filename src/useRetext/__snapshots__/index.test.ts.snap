// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`return creates a disptach object 1`] = `
Object {
  "increment": [Function],
  "sideMenu": Object {
    "child": Object {
      "toggle": [Function],
    },
    "increment": [Function],
    "toggle": [Function],
  },
}
`;

exports[`return keeps the initial state 1`] = `
Object {
  "count": 0,
  "sideMenu": Object {
    "child": Object {
      "isExpanded": false,
    },
    "isOpen": false,
    "maxItems": 100,
  },
}
`;

exports[`store argument throws an error when no action is passed 1`] = `[Error: State, action or reducer are not an object]`;

exports[`store argument throws an error when no object is passed 1`] = `undefined`;

exports[`store argument throws an error when no reducer is passed 1`] = `[Error: State, action or reducer are not an object]`;

exports[`store argument throws an error when no state is passed 1`] = `[Error: State, action or reducer are not an object]`;
